src/Google/GlobalSiteTag.php

```diff
@@ -156,14 +195,241 @@ public function maybe_display_event_snippet( string $ads_conversion_id, string $
+                 "value": "%s",
+                 "currency": "%s",
+                 "transaction_id": "%s"});
+            </script>',
 			esc_js( "{$ads_conversion_id}/{$ads_conversion_label}" ),
 			esc_js( $order->get_total() ),
 			esc_js( $order->get_currency() ),
 			esc_js( $order->get_id() ),
 		);
 	}
 
+	/**
+	 * Display the JavaScript code to track the product view page.
+	 */
+	public function display_view_item_event_snippet(): void {
+		// Only display on the product view page.
+		if ( ! is_product() ) {
+			return;
+		}
+		$product = wc_get_product( get_the_ID() );
+		printf(
+			'<script>
+                gtag("event", "view_item", {
+                    "send_to": "GLA",
+                    "developer_id.%s": "true",
+                    "ecomm_pagetype": "product",
+                    "value": "%s",
+                    items:[{
+                    "id": "gla_%s",
+                    "price": %s,
+                    "google_business_vertical": "retail",
+                    "name":"%s",
+                    "category":"%s",
+                    }]});
+			</script>',
+			esc_js( self::DEVELOPER_ID ),
+			esc_js( (string) $product->get_price() ),
+			esc_js( $product->get_id() ),
+			esc_js( (string) $product->get_price() ),
+			esc_js( $product->get_name() ),
+			esc_js( join( '& ', $product->get_categories() ) ),
+		);
+	}
+
+	/**
+	 * Display the JavaScript code to track all pages.
+	 */
+	public function display_page_view_event_snippet(): void {
+		printf(
+			'<script>
+                gtag(
+                    "event", "page_view", {
+                    "send_to": "GLA",
+                    "developer_id.%s": "true",});
+			</script>',
+			esc_js( self::DEVELOPER_ID )
+		);
+	}
+
+	/**
+	 * Display the JavaScript code to track the cart page.
+	 */
+	public function display_cart_page_snippet(): void {
+		// Only display on the cart page.
+		if ( ! is_cart() ) {
+			return;
+		}
+
+		$item_info = '';
+
+		foreach ( WC()->cart->get_cart() as $cart_item ) {
+			// gets the product id
+			$id = $cart_item['product_id'];
+
+			// gets the product object
+			$product = $cart_item['data'];
+			$name    = $product->get_name();
+			$price   = $product->get_price();
+			// gets the cart item quantity
+			$quantity = $cart_item['quantity'];
+
+			$item_info = $item_info . sprintf(
+				'{
+				"id": "gla_%s",
+				"price": %s,
+				"google_business_vertical": "retail",
+				"name":"%s",
+				"quanitity":"%s",
+				}',
+				esc_js( $id ),
+				esc_js( $price ),
+				esc_js( $name ),
+				esc_js( $quantity )
+			);
+		}
+		$value = WC()->cart->total;
+		printf(
+			'<script>
+            gtag("event", "page_view",
+				{"send_to": "GLA",
+				"ecomm_pagetype": "cart",
+				"value": "%s",
+				 items: [' . esc_html( $item_info ) . ']}); </script>',
+			esc_js( $value ),
+		);
+	}
+
+	/**
+	 * Display the JavaScript code to track the purchase page.
+	 */
+	public function display_purchase_page_snippet(): void {
```

Comment:
Can we combine this with the `maybe_display_event_snippet` and give the function a name that reflects what it's tracking. Combining them means we don't have to run the same conditional checks and load the order twice.

Also the function for tracking the conversion has a safeguard to ensure it's only output once. If we refresh the order received page it won't repeat the conversion event, but this purchase event will be repeated every time the page is releoaded.

Reply:
Thanks for your suggestions.  However, I think it probably couldn't be combined into one function. `maybe_display_event_snippet` is hooked with `wp_head`, but `display_purchase_page_snippet` is hooked with `wp_body_open`. Please correct me if Im wrong.

Reply:
Yes that's correct, they are hooked with two different hooks, but there is no reason they can't use the same one.

In fact I would actually suggest to hook into `woocommerce_before_thankyou` which will already pass the `order_id` so we don't need to retrieve it from the query. I also noticed another flaw, if we look at the WooCommerce thank you (order received) page, it also checks the following condition:

```php
$order->has_status( 'failed' )
```

In both the purchase and conversion events we don't check this condition. So if the payment was denied the order will still be tracked as a conversion/purchase.

Reply:
Thanks for your suggestions. Updated.